    protected void doShutdown() throws Exception {
